docs/RefMan/TypeDeclarations.rst

```diff
@@ -51,3 +51,136 @@ of newtypes to extract the values in the body of the type.
+    then we'll use the first branch of the ``case`` expression and
+    result of the whole expression would be 0;
+
+  * if, ``e`` was create by applying the ``Just`` constructor to some
+    value (e.g, ``Just 2``), then we'll use the second branch of the ``case``
+    expression, and the variable ``a`` will be bound to the value of the field
+    (e.g., ``2``), and the whole expression will evaluate to ``a + 1``
+    (e.g., ``3``).
+
+It is also possible to use just a variable (or ``_``) in a case expression
+to define a catch-all clause---if a value does not match any of the previous
+cases, then this branch will be used:
+
+.. code-block:: cryptol
+
+  isNothing x =
+    case x of
+      Nothing -> True
+      _       -> False
+
+**Upper Case Restriction.**
+The names of the constructors in an ``enum`` declarations
+need to start with an upper-case letter.  This restriction makes it possible
+to distinguish between constructors and variable
+bindings in ``case`` patterns (e.g., between ``Just`` and ``a`` in the
+previous example).
+
+**Non Recursive.** The fields in a constructor may be of any value type,
+as long as this type does not depend on the type to which the constructor
+belongs.  This means that we do not support defining recursive types,
+such as linked lists.
+
+**No Nested Constructor Patterns.**  For simplicity, only non-constructor
+patterns may be used in the fields of a constructor pattern.  For example,
+``Just (a,b)`` and ``Just (a # b)`` are OK, however, ``Just (Just a)``
+will be rejected.  This is a restriction that we may lift in the future.
+
+**No Overlapping Patterns.** For simplicity, all patterns in a
+``case`` expression must be disjoint. In particular, this means that:
+
+  * No two patterns in a ``case`` expression can match the same constructor.
+    This means that Cryptol will reject the following example:
+
+    .. code-block:: cryptol
+
+      isNothing x =
+        case x of
+          Nothing -> True
+          Nothing -> False
+
+  * If a ``case`` expression uses a catch-all clause, then that clause must
+    occur last in the expression. It is an error to match on additional
+    patterns after the catch-all clause. For instance, Cryptol will reject the
+    following example:
+
+    .. code-block:: cryptol
+
+      isNothing x =
+        case x of
+          Just _  -> False
+          _       -> True
+          Nothing -> False
+
+**Patterns Must Be Exhaustive.** The patterns in a ``case`` expression must
+cover all constructors in the ``enum`` type being matched on. For example,
+Cryptol will reject the following example, as it does not cover the ``Just``
+constructor:
+
+.. code-block:: cryptol
+
+  isNothing x =
+    case x of
+      Nothing -> True
+
+**The Matched Expression Must Have a Knwon Enum Type.** Cryptol will reject
+the definition of ``f``, where ``f`` lacks a type signature, or is not not
+an ``enum`` type.
+
+.. code-block:: cryptol
+
+  f x =
+    case x of
+      _ -> ()
```

Comment:
This part reads a bit awkwardly, since it gives two scenarios where a function might be rejected, but the code example only applies to the first scenario. Perhaps we should provide two examples, one for each scenario. I'm thinking something like:

```
**The Matched Expression Must Have a Knwon Enum Type.** Cryptol will reject
the definition of ``f``, where ``f`` lacks a type signature: 

.. code-block:: cryptol
  f x =
    case x of
      _ -> ()

This is because it is not clear... [as before]

Cryptol will also reject this definition, where the type of the value 
being matched is not an ``enum`` type:

.. code-block:: cryptol
  g : Integer -> ()
  g x =
    case x of
      _ -> ()
```